plaso/parsers/gdrive_synclog.py

```diff
@@ -0,0 +1,197 @@
+    message (str): log message.
+    pid (int): process ID of process which logged event.
+    source_code (str): filename:line_number of source file which logged event.
+    thread (str): colon-separated thread identifier in the form "ID:name"
+        which logged event.
+    time (str): date and time of the log entry event with timezone offset.
+  """
+
+  DATA_TYPE = 'gdrive_sync:log:line'
+
+  def __init__(self):
+    """Initializes event data."""
+    super(GoogleDriveSyncLogEventData, self).__init__(data_type=self.DATA_TYPE)
+    self.time = None
+    self.log_level = None
+    self.pid = None
+    self.thread = None
+    self.source_code = None
+    self.message = None
+
+
+class GoogleDriveSyncLogParser(text_parser.PyparsingMultiLineTextParser):
+  """Parses events from Google Drive Sync log files."""
+
+  NAME = 'gdrive_synclog'
+
+  DESCRIPTION = 'Parser for Google Drive Sync log files.'
+
+  _ENCODING = 'utf-8'
+
+  # Increase the buffer size, as log messages are often many lines of python
+  # object dumps or similar.
+  BUFFER_SIZE = 16384
+
+  # Common structures.
+  _HYPHEN = text_parser.PyparsingConstants.HYPHEN
+
+  _FOUR_DIGITS = text_parser.PyparsingConstants.FOUR_DIGITS
+  _THREE_DIGITS = text_parser.PyparsingConstants.THREE_DIGITS
+  _TWO_DIGITS = text_parser.PyparsingConstants.TWO_DIGITS
+
+  # Ignore commas and spaces.
+  IGNORE_FIELD = pyparsing.CharsNotIn(',').suppress()
+
+  _GDS_DATE_TIME = (
+      _FOUR_DIGITS.setResultsName('year') + _HYPHEN +
+      _TWO_DIGITS.setResultsName('month') + _HYPHEN +
+      _TWO_DIGITS.setResultsName('day') +
+      text_parser.PyparsingConstants.TIME_MSEC_ELEMENTS +
+      pyparsing.Word(pyparsing.printables).setResultsName('time_zone_offset')
+  ).setResultsName('date_time')
+
+  # Multiline entry end marker, matched from right to left.
+  _GDS_ENTRY_END = pyparsing.StringEnd() | _GDS_DATE_TIME
+
+  _GDS_LINE = (
+      _GDS_DATE_TIME +
+      pyparsing.Word(pyparsing.alphas).setResultsName('log_level') +
+      # TODO: strip pid= out, cast to integers?
+      pyparsing.Word(pyparsing.printables).setResultsName('pid') +
+      # TODO: consider stripping thread identifier/cleaning up thread name?
+      pyparsing.Word(pyparsing.printables).setResultsName('thread') +
+      pyparsing.Word(pyparsing.printables).setResultsName('source_code') +
+      pyparsing.SkipTo(_GDS_ENTRY_END).setResultsName('message') +
+      pyparsing.ZeroOrMore(pyparsing.lineEnd()))
+
+  LINE_STRUCTURES = [
+      ('logline', _GDS_LINE),
+  ]
+
+  def _ParseLine(self, parser_mediator, structure):
+    """Parses a logline and store appropriate attributes.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      structure (pyparsing.ParseResults): structure of tokens derived from
+          a line of a text file.
+    """
+    # Some hacks follow: our log source uses a different datetime format than
```

Comment:
I had to do a similar approach in https://codereview.appspot.com/332630043/diff/40001/plaso/parsers/sccm.py. Maybe move the date time string creation to a separate method `_GetISO8601String()`

Reply:
Done, though I didn't go as in-depth as sccm.py does with the pre-formatting checks, given Google Drive's log source seems to have less flexibility. I added a couple checks but would prefer not to duplicate checks already in dfdatetime -- LMK if that's foolish.